Move stan/io/json files to the cmdstan repository

rok-cesnovar opened the pull request:
#### Summary:

This PR addresses parts of #719 : 

- creates the directory `cmdstan/src/cmdstan/io/json` and moves all files under `stan/src/stan/io/json` to this created folder

- moves the corresponding tests to the add corresponding unit tests to `src/test/interface/io/json`

- updates cmdstan/src/command.hpp accordingly

After this gets merged, we can add the `RapidJSON` parser lib and replace the existing parser.

#### Intended Effect:

Move JSON IO code to Cmdstan as part of moving to the RapidJSON parser.

#### How to Verify:

Run `python runCmdStanTests.py src/test/interface/io/json/`

#### Side Effects:

/

#### Documentation:

/

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Rok Češnovar, University of Ljubljana

rok-cesnovar commented:
Just a note that Travis tests will not pass due to #728

mitzimorris commented on src/cmdstan/io/json/json_data_handler.hpp:
```diff
@@ -0,0 +1,357 @@
+          errorMsg << "variable: " << key_
+                   << ", error: empty array not allowed";
+          throw json_error(errorMsg.str());
+        }
```
should allow empty array

rok-cesnovar commented:
@mitzimorris thank you very much for the review. Appreciate your time. 

I fixed the behavior on empty arrays and also fixed my previously sloppy job on namespaces (since we are moving stuff to cmdstan we should use cmdstan::json, sorry for missing that before).

mitzimorris commented:
this is ready to merge, but would prefer to wait until we clear up travis problems for sure - @serban-nicusor-toptal, what do you think?

serban-nicusor-toptal commented:
@mitzimorris Hey, I've approved the PR to remove Travis so issue should be resolved now.
Can I help with anything else ?

mitzimorris commented:
many thanks @serban-nicusor-toptal!
then this needs to be merged.
in the meantime, I've done comparison between overall speed of this parser vs. CmdStan 2.20.0 using a model that takes as inputs a few scalar objects and a bunch of vectors and arrays of vectors - data block looks like this:

```
  int<lower=0> N;
  int<lower=0> N_edges;
  int<lower=1, upper=N> node1[N_edges];  // node1[i], node2[i] neighbors
  int<lower=1, upper=N> node2[N_edges];  // node1[i] < node2[i]

  int<lower=0> y[N];             // count outcomes
  vector<lower=0>[N] E;          // exposure
  int<lower=1> K;                // num covariates
  matrix[N, K] x;                // design matrix
  real<lower=0> scaling_factor;  // scales the variance of the spatial effects
```

testing with dataset where N = 2095, and one where N = 20950 -
total number of chars in each file:  151,159   and 1,035,484  respectively.

model itself does (almost) nothing:

```
parameters {
  real beta0;            // intercept
}
model {
  beta0 ~ normal(0, 1);
}
```

compiled model and tested by running 10 times under both CmdStan 2.20.0 and this branch -
```
for i in {1..10}; do echo $i; time (./bym2_read_data sample num_warmup=1 num_samples=1 data file=bym2_big.data.json >/dev/null); echo ""; done
```

this minimizes IO for writing output.csv file - the bulk of the work should be reading in the data file, correct?  the timings were pretty much the same for both versions.

to me this indicates that parsing the JSON is just not a big deal compared to the work done by the model during model instantiation - this is what the model constructor does with that var_context object:

```
            current_statement_begin__ = 7;
            validate_non_negative_index("y", "N", N);
            context__.validate_dims("data initialization", "y", "int", context__.to_vec(N));
            y = std::vector<int>(N, int(0));
            vals_i__ = context__.vals_i("y");
            pos__ = 0;
            size_t y_k_0_max__ = N;
            for (size_t k_0__ = 0; k_0__ < y_k_0_max__; ++k_0__) {
                y[k_0__] = vals_i__[pos__++];
            }
            size_t y_i_0_max__ = N;
            for (size_t i_0__ = 0; i_0__ < y_i_0_max__; ++i_0__) {
                check_greater_or_equal(function__, "y[i_0__]", y[i_0__], 0);
            }

            current_statement_begin__ = 8;
            validate_non_negative_index("E", "N", N);
            context__.validate_dims("data initialization", "E", "vector_d", context__.to_vec(N));
            E = Eigen::Matrix<double, Eigen::Dynamic, 1>(N);
            vals_r__ = context__.vals_r("E");
            pos__ = 0;
            size_t E_j_1_max__ = N;
            for (size_t j_1__ = 0; j_1__ < E_j_1_max__; ++j_1__) {
                E(j_1__) = vals_r__[pos__++];
            }
            check_greater_or_equal(function__, "E", E, 0);

            current_statement_begin__ = 9;
            context__.validate_dims("data initialization", "K", "int", context__.to_vec());
            K = int(0);
            vals_i__ = context__.vals_i("K");
            pos__ = 0;
            K = vals_i__[pos__++];
            check_greater_or_equal(function__, "K", K, 1);

            current_statement_begin__ = 10;
            validate_non_negative_index("x", "N", N);
            validate_non_negative_index("x", "K", K);
            context__.validate_dims("data initialization", "x", "matrix_d", context__.to_vec(N,K));
            x = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>(N, K);
            vals_r__ = context__.vals_r("x");
            pos__ = 0;
            size_t x_j_2_max__ = K;
            size_t x_j_1_max__ = N;
            for (size_t j_2__ = 0; j_2__ < x_j_2_max__; ++j_2__) {
                for (size_t j_1__ = 0; j_1__ < x_j_1_max__; ++j_1__) {
                    x(j_1__, j_2__) = vals_r__[pos__++];
                }
            }
```

seantalts commented:
@mitzimorris could you try running that setup within Apple's Instruments (assuming you're on a Mac) and profiling it? I'd be curious to see what it claims the biggest blocks of time are spent doing. It's pretty disheartening that the times are similar; this time we should profile so we aim ourselves at the actual problem instead of assuming again.

rok-cesnovar commented:
This is still the same existing parser. When we decided we want json parsing in cmdstan I [mentioned in the issue](https://github.com/stan-dev/cmdstan/issues/719#issuecomment-524975860) that I would want to do it in 2 phases: 
- move the existing parser to cmdstan and remove the files from stan (ready after this is merged)
- remove the existing parser with rapidjson

I wanted to do it this way because if we find anything wrong with rapidjson after it gets merged, reverting will be easier.

rok-cesnovar commented:
I am sorry if that isnt clear from the PR text :/ 

Will definitely do the performance test and make a script so anyone can verify once I open that PR (hopefully today). I am going to use the optimize example from [here](https://jeffpollock9.github.io/maximum-likelihood-estimation-with-tensorflow-probability-and-stan-take-2/)

mitzimorris commented:
so glad to hear it!
would you like the model and data files for this test I just ran?

rok-cesnovar commented:
Yes, please!

seantalts commented:
> This is still the same existing parser. When we decided we want json parsing in cmdstan I mentioned in the issue that I would want to do it in 2 phases:

Phew!

mitzimorris commented:
just added the files to src/test/test-models for this branch.
